Home Work № 5

skyseedim opened the pull request.

skyseedim committed:
Home Work № 5

NicknameSoHard commented on HW5.py:
```diff
@@ -0,0 +1,168 @@
+print("\nTask № 2")
+with open("Task_2_text.txt") as f_obj:
+    cont = f_obj.readlines()
+res = []
+res.extend([(i+1, len(string.split())) for i, string in enumerate(cont)])
```
Так в общем то сразу бы записали `res= [(i+1, len(string.split())) for i, string in enumerate(cont)]`, а то какие-то лишние действия происходят.

NicknameSoHard commented on HW5.py:
```diff
@@ -0,0 +1,168 @@
+
+res = []
+res.extend([[string.split("-")[0].rstrip(), string.split("-")[1]] for string in cont])
+res[0][0] = "Один -"
+res[1][0] = "Два -"
+res[2][0] = "Три -"
+res[3][0] = "Четыре -"
```
А если в файле изменится порядок, то что будете весь код переписывать? Или если добавятся еще слова для замены? Не слишком то удобно получается.
Может быть лучше подмену слов реализовать через какую-то структуру данных, которая по одному слову может выдавать какое-то другое значение? Вроде словаря. Будет более универсально и легко расширяемо.

NicknameSoHard commented on HW5.py:
```diff
@@ -0,0 +1,168 @@
+res[2][0] = "Три -"
+res[3][0] = "Четыре -"
+with open("Task_4_text_2.txt", "w") as f_obj:
+    [f_obj.writelines(string) for string in res]
```
`writelineS` рассчитана на то чтобы туда подать список значений для записи. Поэтому для использования этой функции лучше писать так:
`f_obj.writelines([string for string in res])`, т.е. сформируете список из строчек что нужно записать, а затем передадите этот список для записи в функцию. Так получится что вы сформируете набор значений и всего раз его запишете.

(Но на самом деле по идее генераторы нужны чтобы сформировать какое-то значение или отфильтровать, поэтому лучше их использовать так чтобы их значение куда-то дальше уходило, вроде как в первом варианте, поэтому он предпочтительнее)

NicknameSoHard commented on HW5.py:
```diff
@@ -0,0 +1,168 @@
+with open("Task_7_text.txt") as f_obj:
+    cont = f_obj.readlines()
+# переводим данные из файла в словарь - название фирмы (ключ) и рассчитанная прибыль или убыток
+my_dict = {les.split()[0]: float(les.split()[2]) - float(les.split()[3]) for les in cont}
```
Я бы рекомендовал расписать это формирование не генератором, так как технически вы каждую строку трижды разделяете, что не слишком то эффективно, хоть и коротко. Просто не очень оптимально выходит, хоть и результат тот же.

skyseedim committed:
Изменения в решениях задач 2, 4, 7

Задание 2:  удалено  создание пустого списка res и использование ф. extend для него, теперь генератор сразу  подаётся в res

Задание 4:
-  удалено  создание пустого списка res и использование ф. extend для него, теперь генератор сразу  подаётся в res
- полностью изменено решение, теперь создаётся словарь с переводами чисел, замена слов происходит в цикле for в зависимости от числа строк в изначальном файле, теперь неважен порядок чисел в файле.

Задание 7: теперь, когда переводим нужные данные из файла в словарь, добавлено  формирование списка всех данных из файла через функцию split, а уже потом формируем словарь с нужными данными из нового списка. Таким образом, split применяется один раз для каждой строчки (а не три).
